Emit a warning when 'load_from_checkpoint' called from instance.

kwsp opened the pull request:
## What does this PR do?

<!--
Please include a summary of the change and which issue is fixed.
Please also include relevant motivation and context.
List any dependencies that are required for this change.

If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

The following links the related issue to the PR (https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
-->

Fixes https://github.com/Lightning-AI/lightning/issues/18169

<!-- Does your PR introduce any breaking changes? If yes, please list them. -->

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
  - Updated docstring
- Did you write any **new necessary tests**? (not for typos and docs)
  - I'm not away of a way to test for the warning we emit.
- [x ] Did you verify new and **existing tests pass** locally with your changes?
  - Tests were broken on master on my machine prior to my changes.
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)


<!-- In the CHANGELOG, separate each item in the unreleased section by a blank line to reduce collisions -->

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

<!--

Did you have fun?

Make sure you had fun coding 🙃

-->

kwsp commented:
`restricted_classmethod` might be a good fit for `lightning_utilities`, so we could land a PR there first.

carmocca commented on src/lightning/pytorch/utilities/restricted_classmethod.py:
```diff
@@ -0,0 +1,19 @@
+        if instance is not None:
+            rank_zero_warn(
+                f"The classmethod {cls.__name__}.{self.method.__name__} was called on an instance. Please "
+                f"call this method on the class type ({cls.__name__}) and make sure the return value is used"
```
Should it instead raise an error?

Borda replied:
I would say, warn and return nothing or raise an error

kwsp replied:
raising an error would technically be backwards incompatible if there are valid user code that calls the method on an instance but actually use the return value.

carmocca replied:
My question then is: which would be those valid usecases?

carmocca commented on src/lightning/pytorch/utilities/restricted_classmethod.py:
```diff
@@ -0,0 +1,19 @@
+from lightning_utilities.core.rank_zero import rank_zero_warn
```
```suggestion
from lighting.pytorch.utilities.rank_zero import rank_zero_warn
```

carmocca commented on src/lightning/pytorch/utilities/restricted_classmethod.py:
```diff
@@ -0,0 +1,19 @@
+    called on an instance and not the class type.
+    """
+
+    def __init__(self, method):
```
Can you add typing annotations?

carmocca commented on src/lightning/pytorch/utilities/restricted_classmethod.py:
```diff
@@ -0,0 +1,19 @@
+                f"The classmethod {cls.__name__}.{self.method.__name__} was called on an instance. Please "
+                f"call this method on the class type ({cls.__name__}) and make sure the return value is used"
+            )
+        return lambda *args, **kwargs: self.method(cls, *args, **kwargs)
```
I believe this will work too

```suggestion
        return self.method.__get__(cls, cls)
```

carmocca commented:
I would land this here first and upstream it later, otherwise this will be blocked by releasing a new lightning-utilities version and increasing the minimum dependency so that we can use it

kwsp commented:
Changed from warning to exception per discussion. Wrote a new test for `LightningModule.load_from_checkpoint`. However, I found no existing tests for `LightningDataModule.load_from_checkpoint` so I didn't add a test for that (as I don't know how the infrastructure would work there).

carmocca commented on src/lightning/pytorch/utilities/restricted_classmethod.py:
```diff
@@ -0,0 +1,31 @@
+
+    def __get__(self, instance: _T, cls: type[_T]) -> Callable[_P, _R_co]:
+        if instance is not None:
+            raise RestrictedClassmethodError(
```
Should we simply use TypeError? The type of the input is not the expected

carmocca commented on tests/tests_pytorch/core/test_saving.py:
```diff
@@ -116,3 +116,17 @@ def test_load_from_checkpoint_warn_on_empty_state_dict(tmp_path):
     assert model.device.type == "cpu"
+
+
+def test_load_from_checkpoint_classmethod(tmp_path):
```
Can you instead test the restricted checkpoint by creating a class with a classmethod that sets it? We don't need to create a checkpoint and load a model to test it

carmocca commented on src/lightning/pytorch/utilities/restricted_classmethod.py:
```diff
@@ -0,0 +1,31 @@
+    """
+
+
+_T = TypeVar("_T")
+_P = ParamSpec("_P")
+_R_co = TypeVar("_R_co", covariant=True)
```
This level of typing is more complex than what the average code reader would understand. Mind adding some comments that explain why we need all these?

awaelchli commented on src/lightning/pytorch/utilities/restricted_classmethod.py:
```diff
@@ -0,0 +1,23 @@
+
+class restricted_classmethod(Generic[_T, _P, _R_co]):
+    """
+    Custom `classmethod` that emits a warning when the classmethod is
```
```suggestion
    Custom `classmethod` that raises an exception when the classmethod is
```

awaelchli commented on src/lightning/pytorch/utilities/restricted_classmethod.py:
```diff
@@ -0,0 +1,23 @@
+    def __get__(self, instance: _T, cls: type[_T]) -> Callable[_P, _R_co]:
+        if instance is not None:
+            raise TypeError(
+                f"The classmethod `{cls.__name__}.{self.method.__name__}` cannot be invoked on an instance. Please "
+                f"invoke it on the class type and make sure the return value is used."
```
I propose simpler language
```suggestion
                f"The classmethod `{cls.__name__}.{self.method.__name__}` cannot be called on an instance. Please "
                f"call it on the class type and make sure the return value is used."
```

awaelchli commented on tests/tests_pytorch/utilities/test_restricted_classmethod.py:
```diff
@@ -0,0 +1,18 @@
+
+class _C:
+    @restricted_classmethod
+    def _rcmethod(cls):
```
```suggestion
    def restricted_method(cls):
```

awaelchli commented on src/lightning/pytorch/utilities/restricted_classmethod.py:
```diff
@@ -0,0 +1,23 @@
+_R_co = TypeVar("_R_co", covariant=True)  # return type of the decorated method
+
+
+class restricted_classmethod(Generic[_T, _P, _R_co]):
```
I suggest to make this protected so we can change it in the future if needed. 

```suggestion
class _restricted_classmethod(Generic[_T, _P, _R_co]):
```

awaelchli reviewed (commented):
Thanks for working on this!

kwsp commented:
@awaelchli I addressed your review.

kwsp commented:
Don't merge yet- I just learned a better way to type _restricted_classmethod to make static type checkers happy. I'll commit later when I get to a computer.

kwsp commented:
@carmocca I fixed the static typing problem and removed the `# type: ignore` directives after discussion here: https://github.com/microsoft/pyright/issues/5865

awaelchli commented:
@kwsp Not sure if you can see the failing tests since you are a first-time contributor, so I'm pasting one failure here:

```
______________ test_tensorboard_log_graph[example_input_array1] _______________

tmpdir = local('C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pytest-of-runneradmin\\pytest-0\\test_tensorboard_log_graph_exa0')
example_input_array = tensor([[0.7449, 0.2476, 0.4412, 0.8661, 0.9531, 0.6571, 0.1275, 0.8914, 0.3353,
         0.4487, 0.1779, 0.4472, 0.68...571, 0.0477, 0.0973, 0.6662, 0.9150, 0.1149, 0.8697, 0.7949, 0.4071,
         0.0140, 0.7669, 0.1640, 0.5071, 0.2030]])

    @pytest.mark.parametrize("example_input_array", [None, torch.rand(2, 32)])
    def test_tensorboard_log_graph(tmpdir, example_input_array):
        """Test that log graph works with both model.example_input_array and if array is passed externally."""
        model = BoringModel()
        if example_input_array is not None:
            model.example_input_array = None
    
        logger = TensorBoardLogger(tmpdir, log_graph=True)
>       logger.log_graph(model, example_input_array)

D:\a\lightning\lightning\tests\tests_pytorch\loggers\test_tensorboard.py:219: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\lightning_utilities\core\rank_zero.py:32: in wrapped_fn
    return fn(*args, **kwargs)
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\lightning\pytorch\loggers\tensorboard.py:215: in log_graph
    self.experiment.add_graph(model, input_array)
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\torch\utils\tensorboard\writer.py:841: in add_graph
    graph(model, input_to_model, verbose, use_strict_trace)
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\torch\utils\tensorboard\_pytorch_graph.py:337: in graph
    trace = torch.jit.trace(model, args, strict=use_strict_trace)
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\torch\jit\_trace.py:759: in trace
    return trace_module(
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\torch\jit\_trace.py:960: in trace_module
    module = make_module(mod, _module_class, _compilation_unit)
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\torch\jit\_trace.py:574: in make_module
    elif torch._jit_internal.module_has_exports(mod):
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\torch\_jit_internal.py:749: in module_has_exports
    if hasattr(mod, name):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <lightning.pytorch.utilities.restricted_classmethod._restricted_classmethod_impl object at 0x0000028804C3E7D0>
instance = BoringModel(
  (layer): Linear(in_features=32, out_features=2, bias=True)
)
cls = <class 'lightning.pytorch.demos.boring_classes.BoringModel'>

    def __get__(self, instance: _T | None, cls: type[_T]) -> Callable[_P, _R_co]:
        if instance is not None:
>           raise TypeError(
                f"The classmethod `{cls.__name__}.{self.method.__name__}` cannot be called on an instance. "
                f"Please call it on the class type and make sure the return value is used."
            )
E           TypeError: The classmethod `BoringModel.load_from_checkpoint` cannot be called on an instance. Please call it on the class type and make sure the return value is used.

```

The newly added error in `__get__` gets triggered when tracing a module with `torch.jit.trace()` which happens in some tests (even if no `.load_from_checkpoint()` call occurs. There is probably a way to detect the tracing to fix this issue.

kwsp commented:
@awaelchli I see. I'll investigate later how to make it play nice with `torch.jit.trace`.

carmocca commented:
Uh, that might be a problem. Detecting it doesnt always work. In fact, we have this hack just because of it: https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/core/module.py#L1590-L1597

awaelchli commented on src/lightning/pytorch/utilities/restricted_classmethod.py:
```diff
@@ -0,0 +1,34 @@
+        self.method = method
+
+    def __get__(self, instance: _T | None, cls: type[_T]) -> Callable[_P, _R_co]:
+        if instance is not None:
+            raise TypeError(
+                f"The classmethod `{cls.__name__}.{self.method.__name__}` cannot be called on an instance. "
+                f"Please call it on the class type and make sure the return value is used."
+            )
```
Hey @carmocca @kwsp I think this is an important PR because this mistake happens very frequently. I would suggest we put the jit scripting at a lower priority and compromise on making a hack to enable this PR. I played around with the following check and it seems to work where `torch.is_jit_scripting()` fails. 

```suggestion
        if instance is not None and not any("torch/jit" in s.filename for s in inspect.stack()):
            raise TypeError(
                f"The classmethod `{cls.__name__}.{self.method.__name__}` cannot be called on an instance. "
                f"Please call it on the class type and make sure the return value is used."
            )
```

Should we try this?

carmocca replied:
Fine with me if it works. We should evaluate removing support for `torch.jit` soon.

kwsp commented:
The mypy CI failing is because currently no static type checker understands custom classmethods like the hack we implemented. Might just have to do `# type: ignore [arg-type]` there

awaelchli commented on src/lightning/pytorch/core/datamodule.py:
```diff
@@ -157,7 +158,7 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
         pass
 
-    @classmethod
+    @_restricted_classmethod
```
Is the Self return annotation still correct now that we've changed the decorator?
```
src/lightning/pytorch/core/module.py:1550: error: Argument 1 to "_load_from_checkpoint" has incompatible type "Self"; expected "type[LightningModule] | type[LightningDataModule]"  [arg-type]
src/lightning/pytorch/core/datamodule.py:228: error: Argument 1 to "_load_from_checkpoint" has incompatible type "Self"; expected "type[LightningModule] | type[LightningDataModule]"  [arg-type]

```

Should we ignore the typing error or change the type?

carmocca replied:
Ignore it IMO

awaelchli reviewed (approved):
Tests pass now. The classmethod wrapper seems to work well, thanks! This resolves a long-standing usability issue 🎉
